### PR TITLE
Add regular & fixme tests for function contracts

### PR DESCRIPTION
**Pull request by celinval**
I'm adding a few fix-me tests that I bumped into while working on #3363. Most of them will be fixed by #3363, except the one related to #3370.

The original PR is already quite large, so I decided to just push all of these as fixme tests for now. This is now ready for review!

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: tests/kani/FunctionContracts/fixme_receiver_contracts.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+#![feature(rustc_attrs)]
+
+extern crate kani;
+
+use std::boxed::Box;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::sync::Arc;
+
+/// Type representing a valid ASCII value going from `0..=128`.
```
Comment (zhassan-aws) [inline tests/kani/FunctionContracts/fixme_receiver_contracts.rs:23]: Should it be:
```suggestion
/// Type representing a valid ASCII value going from `0..128`.
```
?

File: tests/kani/FunctionContracts/fixme_receiver_contracts.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+mod verify {
+    use super::*;
+    use kani::Arbitrary;
+
+    #[kani::proof_for_contract(CharASCII::set_val)]
+    fn check_set_val() {
+        let mut obj = CharASCII::any();
+        let original = obj.0;
+        let new_val = kani::any_where(|new| *new != original);
```
Comment (zhassan-aws) [inline tests/kani/FunctionContracts/fixme_receiver_contracts.rs:97]: Is this needed?

File: tests/kani/FunctionContracts/fixme_old_requires_order.rs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! Check that `old()` is executed after the pre-conditions, otherwise it can fail.
+//!
+//! Issue: <https://github.com/model-checking/kani/issues/3370>
+// kani-flags: -Zfunction-contracts
+
+#[kani::requires(val < i32::MAX)]
+#[kani::ensures(|result| *result == old(val + 1))]
```
Comment (zhassan-aws) [inline tests/kani/FunctionContracts/fixme_old_requires_order.rs:9]: Should it be:
```suggestion
#[kani::ensures(|result| *result == old(val) + 1)]
```
?

File: tests/kani/FunctionContracts/fixme_old_requires_order.rs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+pub fn next(mut val: i32) -> i32 {
+    val + 1
+}
+
+#[kani::proof_for_contract(next)]
+pub fn check_next() {
+    // let _ = next(kani::any_where(|val: &i32| *val < i32::MAX));
```
Comment (zhassan-aws) [inline tests/kani/FunctionContracts/fixme_old_requires_order.rs:16]: Remove commented line.

File: tests/kani/FunctionContracts/fixme_old_requires_order.rs (hunk 3 above)
Comment (celinval) [reply tests/kani/FunctionContracts/fixme_old_requires_order.rs:9]: it doesn't matter...

File: tests/kani/FunctionContracts/fixme_receiver_contracts.rs (hunk 2 above)
Comment (celinval) [reply tests/kani/FunctionContracts/fixme_receiver_contracts.rs:97]: No.. I used to have an assertion after that relied on it.

File: tests/kani/FunctionContracts/fixme_receiver_contracts.rs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+/// Type representing a valid ASCII value going from `0..128`.
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[rustc_layout_scalar_valid_range_start(0)]
+#[rustc_layout_scalar_valid_range_end(128)]
+struct CharASCII(u8);
+
+impl kani::Arbitrary for CharASCII {
+    fn any() -> CharASCII {
+        let val = kani::any_where(|inner: &u8| *inner <= 128);
```
Comment (zhassan-aws) [inline tests/kani/FunctionContracts/fixme_receiver_contracts.rs:31]: This now needs to be updated to a strict less than (`*inner < 128`).

File: tests/kani/FunctionContracts/fixme_receiver_contracts.rs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        unsafe { CharASCII(val) }
+    }
+}
+
+/// This type contains unsafe setter functions with the same contract but different type of
+/// receivers.
+impl CharASCII {
+    #[kani::modifies(&self.0)]
+    #[kani::requires(new_val <= 128)]
```
Comment (zhassan-aws) [inline tests/kani/FunctionContracts/fixme_receiver_contracts.rs:40]: Same here (and other occurrences below).

File: tests/kani/FunctionContracts/fixme_receiver_contracts.rs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        let mut obj = CharASCII::any();
+        let original = obj.0;
+        let new_val: u8 = kani::any();
+        unsafe { obj.set_val(new_val) };
+    }
+
+    #[kani::proof_for_contract(CharASCII::set_mut_ref)]
+    fn check_mut_ref() {
+        let mut obj = CharASCII::any();
+        let original = obj.0;
```
Comment (zhassan-aws) [inline tests/kani/FunctionContracts/fixme_receiver_contracts.rs:104]: Is this line serving a purpose?

File: tests/kani/FunctionContracts/modify_slice_elem.rs [hunk 8]
```diff
@@ -0,0 +1,25 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! Check that Kani correctly verify the contract that modifies slices.
```
Comment (zhassan-aws) [inline tests/kani/FunctionContracts/modify_slice_elem.rs:3]:
```suggestion
//! Check that Kani correctly verifies the contract that modifies slices.
```

**Review by zhassan-aws — APPROVED** [review]

File: charon
[the remark this answers is not in the document]
Comment (celinval) [reply charon:1]: Merging main fixed this.

File: tests/kani/FunctionContracts/fixme_receiver_contracts.rs (hunk 7 above)
Comment (celinval) [reply tests/kani/FunctionContracts/fixme_receiver_contracts.rs:104]: I guess not...
```suggestion
```

**Pull request merged**